### PR TITLE
fix: fastify adapter using fastify-plugin

### LINKED ISSUE #1715: Fastify adapter: Custom content-type parser for application/json is breaking normal REST api requests on a fastify server
https://github.com/trpc/trpc/blob/d5808b863490f95ff76defce4886833b11672e88/packages/server/src/adapters/fastify/fastifyTRPCPlugin.ts#L25-L31

This works just fine for tRPC endpoints but breaks other REST api endpoints as `request.body` is a string where it's expected to an object. This can be easily reproduced on the official fastify examples by creating a REST `POST` endpoint and sending some data inside body.

### PR DESCRIPTION
**Pull request by skarab42**
Fixes #1715, allows the use of the fastify adapter with or without the `fastify-plugin`.

With fp: the response body will be a string for all requests (as before). 
Without fp: the response body will not be modified by the adapter.

### CODE REVIEW

**Comment by soulsam480** [comment]
nice work @skarab42 :blue_heart:

File: examples/fastify-server/src/client/index.ts [hunk 1]
```diff
@@ -10,6 +10,23 @@ async function start() {
     headers: { username: 'nyan' },
   });
 
+  const getHello = await fetch(`http://localhost:${port}/hello`);
+  const getHelloJSON = await getHello.json();
+
+  console.log('>>> fetch:get:hello:', getHelloJSON);
```
Comment (jlalmes) [inline examples/fastify-server/src/client/index.ts:16]: Remove the `console.log`s in this file?

Comment (jlalmes) [reply examples/fastify-server/src/client/index.ts:16]: Oh whoops didn't realise this file is in the `examples/**` folder, please ignore me!

**Review by KATT — COMMENTED** [review]
Just a quick check - Is this a breaking change for existing usage?

File: packages/server/src/adapters/fastify/fastifyTRPCPlugin.ts [hunk 2]
```diff
@@ -30,7 +30,15 @@ export function fastifyTRPCPlugin<TRouter extends AnyRouter>(
...
   );
 
-  fastify.all(`${opts.prefix ?? ''}/:path`, async (req, res) => {
+  let prefix = opts.prefix ?? '';
+
+  // https://github.com/fastify/fastify-plugin/blob/fe079bef6557a83794bf437e14b9b9edb8a74104/plugin.js#L11
+  // @ts-expect-error property 'default' does not exists on type ...
+  if (typeof fastifyTRPCPlugin.default !== 'function') {
+    prefix = ''; // handled by fastify internally
+  }
```
[the remark this answers is not in the document]
Comment (skarab42) [reply packages/server/src/adapters/fastify/fastifyTRPCPlugin.ts:39]: It is declared in local scope.
https://github.com/trpc/trpc/blob/e045359f64f3cf6b16f8bdd0f3cae51b8a13fa68/packages/server/src/adapters/fastify/fastifyTRPCPlugin.ts#L20

[the remark this answers is not in the document]
Comment (sachinraja) [reply packages/server/src/adapters/fastify/fastifyTRPCPlugin.ts:39]: Looking at the link in the comment it says this:
> Support for 'export default' behaviour in transpiled ECMAScript module
    fn = fn.default

Our bundling setup is not the same and this hack only applies to `export default`, not a named `export`

**Comment by skarab42** [comment]
@KATT @sachinraja What is strange is the [error message in the CI](https://github.com/trpc/trpc/runs/5777989927?check_suite_focus=true#step:13:16). The `wait-on` command says that the connection timeout on port 2022 has been reached.  But the [previous line](https://github.com/trpc/trpc/runs/5777989927?check_suite_focus=true#step:13:15) shows that the server is up on the right port.

File: packages/server/src/adapters/fastify/fastifyTRPCPlugin.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (skarab42) [reply packages/server/src/adapters/fastify/fastifyTRPCPlugin.ts:39]: Ok I see, but it [does in any case](https://github.com/fastify/fastify-plugin/blob/fe079bef6557a83794bf437e14b9b9edb8a74104/plugin.js#L52) and bundlers do not remove this property. I haven't found any other way to know if the module is wrapped by `fastify-plugin`, If you have a cleaner solution, I'd like to see it.

**Comment by skarab42** [comment]
Fix fixed 🚀

**Review by KATT — APPROVED** [review]

**Pull request merged**